src/cplscheme/CompositionalCouplingScheme.cpp

```diff
@@ -15,96 +18,114 @@ void CompositionalCouplingScheme::addCouplingScheme(
   }
-  determineActiveCouplingSchemes();
 }
 
 void CompositionalCouplingScheme::receiveResultOfFirstAdvance()
 {
-  for (const Scheme &scheme : _couplingSchemes) {
-    scheme.scheme->receiveResultOfFirstAdvance();
+  for (const auto scheme : allSchemes()) {
+    scheme->receiveResultOfFirstAdvance();
   }
 }
 
 bool CompositionalCouplingScheme::sendsInitializedData() const
 {
   PRECICE_TRACE();
-  bool sendsInitializedData = false;
-  for (const Scheme &scheme : _couplingSchemes) {
-    sendsInitializedData |= scheme.scheme->sendsInitializedData();
-  }
+  auto schemes              = allSchemes();
+  bool sendsInitializedData = std::any_of(schemes.begin(), schemes.end(), std::mem_fn(&CouplingScheme::sendsInitializedData));
   PRECICE_DEBUG("return {}", sendsInitializedData);
   return sendsInitializedData;
 }
 
 bool CompositionalCouplingScheme::isInitialized() const
 {
   PRECICE_TRACE();
-  bool isInitialized = true;
-  for (const Scheme &scheme : _couplingSchemes) {
-    isInitialized &= scheme.scheme->isInitialized();
-  }
-  PRECICE_DEBUG("return {}", isInitialized);
+  auto schemes       = allSchemes();
+  bool isInitialized = std::any_of(schemes.begin(), schemes.end(), std::mem_fn(&CouplingScheme::isInitialized));
   return isInitialized;
 }
 
 void CompositionalCouplingScheme::addComputedTime(double timeToAdd)
 {
   PRECICE_TRACE(timeToAdd);
-  for (SchemesIt it = _activeSchemesBegin; it != _activeSchemesEnd; it++) {
-    if (not it->onHold) {
-      it->scheme->addComputedTime(timeToAdd);
-    }
+  for (const auto scheme : schemesToRun()) {
+    scheme->addComputedTime(timeToAdd);
   }
-  _lastAddedTime += timeToAdd;
 }
 
-void CompositionalCouplingScheme::advance()
+CouplingScheme::ChangedMeshes CompositionalCouplingScheme::firstSynchronization(const CouplingScheme::ChangedMeshes &changes)
 {
   PRECICE_TRACE();
-  bool moreSchemesToHandle = false;
-  do {
-    for (SchemesIt it = _activeSchemesBegin; it != _activeSchemesEnd; it++) {
-      if (not it->onHold) {
-        it->scheme->advance();
-      }
-    }
-    moreSchemesToHandle = determineActiveCouplingSchemes();
-    if (moreSchemesToHandle) {
-      // The new schemes to be handled in this advance also need the time that
-      // has been computed so far. This time can't be added in the solver call
-      // to addComputedTime(), since there the schemes are not active yet.
-      addComputedTime(_lastAddedTime);
-    }
-  } while (moreSchemesToHandle);
-  _lastAddedTime = 0.0;
+  PRECICE_ASSERT(changes.empty());
+  CouplingScheme::ChangedMeshes totalChanges;
+  for (const auto scheme : schemesToRun()) {
+    auto remoteChanges = scheme->firstSynchronization(changes);
+    totalChanges.insert(totalChanges.end(), remoteChanges.begin(), remoteChanges.end());
+  }
+  return totalChanges;
+}
+
+void CompositionalCouplingScheme::firstExchange()
+{
+  PRECICE_TRACE();
+  for (const auto scheme : schemesToRun()) {
+    scheme->firstExchange();
```

Comment:
Maybe a matter of taste but we could use `for_each` here and in several other places.